src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -555,16 +557,34 @@ protected synchronized int engineDoFinal(byte[] bytes, int offset, int length, b
             }
 
             if (opMode == NATIVE_MODE_DECRYPT) {
-                // We use this temporary buffer both for our input ciphertext and our output (possibly unauthenticated) plaintext.
-                // We'll copy it to the actual output location iff decryption completes successfully.
-                engineUpdate(bytes, offset, length); // Decrypt mode never generates output for updates
-                final byte[] tempBuffer = decryptInputBuf.getDataBuffer();
-                final int tempBufferLength = decryptInputBuf.size();
-
-                if (tempBufferLength < tagLength) {
+                // We cannot write directly to the output array because that could release
+                // possibly unauthenticated plaintext.
+                // If we have already processed some amount of ciphertext, then we can use our existing input buffer
+                // for both input and output and then copy it over to the real output upon success.
+                // If we have not already processed some ciphertext, then we can use the input buffer as is
+                // and allocate temporary output buffer.
+                if (decryptInputBuf.size() + length < tagLength) {
```

Comment:
isn't this inconsistent with [how we validate tagLen in our C++ shim](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/csrc/aes_gcm.cpp#L395)?

Reply:
No, this is the same logic. However, this is prior to us merging data in `decryptInputBuf` and `bytes` (whose length is represented by `length`). So, after we merge the two, `inlen` (in C++) is equal to `decryptInputBuf.size() + length` (in Java).

Reply:
ah got it. i think the root cause of my misunderstanding here was also missing the fact that L576's call to `engineUpdate` [modifies `decryptInputBuf`](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/AesGcmSpi.java#L425) and increases its size.